### DISCUSSION TITLE
Can't get `fitView` to work at all...

### DISCUSSION THREAD

**Discussion by CSworenI** [xyflow/xyflow#4575] [Q&A] [answered]
I have tried every way I can find and it's not working for me. I am trying to time it with the opening/closing of an options pane (`aside` element) with the changing of a state (which also controls the viewable space of the grid). I have tried it multiple ways:

With the `onClick` event:
```
onClick={() => [setOptions(false), fitView()]}
```

`useEffect` triggered on the state as described above:
```
	useEffect(() => {
		fitView();
	}, [options]);
```

similar, but per documentation (may be old?):
```
	useEffect(() => {
		window.requestAnimationFrame(() => fitView());
	}, [options]);
```

and another similar, but using the instance of the `ReactFlow` component:
```
	useEffect(() => {
		window.requestAnimationFrame(() => flowInstance.fitView());
	}, [options]);
```

I've also tried each combination without the `window.requestAnimationFrame` clause. None of these seem to work. Can you advise?

**Comment by Santoshsohani** [comment]
```js
   <ReactFlow
        nodes={nodes}
        edges={edges}
        onNodesChange={onNodesChange}
        onEdgesChange={onEdgesChange}
        onConnect={onConnect}
        fitView
      >
        <MiniMap />
        <Controls />
        <Background variant="dots" />
        <Panel position="top-right">
          <button onClick={handleAddNode} className='bg-purple-800 text-white font-bold py-2 px-4 rounded'>Add Node</button>
        </Panel>
   </ReactFlow>
```

Have you tried `fitview` as a prop for `<ReactFlow>`

**Reply by moklick** [reply]
Hey @CSworenI 

if you want to add a click handler, you would need to use curly brackets:

```js
onClick={() => {
  setOptions(false);
  fitView();
}
```

If you want to use it in an `useEffect`, you would need to pass `fitView` as a dependency:

```js
useEffect(() => {
  fitView();
}, [options, fitView]);
```

**Reply by CSworenI** [reply]
@Santoshsohani Yes, `fitView` is an attribute of my `ReactFlow` component and has been.

@moklick I have tried both of those and still don't get the `fitView` applied for the `onClick` event or with `useEffect`. I even put a console log in each and I'm getting those logs, but the view isn't fitting to the resized window. I tried with both the `useReactFlow` and flow instance versions of `fitView` without success.

**Reply by CSworenI** [reply, marked as answer]
Ahhh, but adding a `setTimeout` with at least 5ms works. So apparently it _was_ working, but needed the delay to catch the change of state. So fun, this React learning for me:
```
	useEffect(() => {
		setTimeout(() => flowInstance.fitView(), 5);
	}, [options, fitView]);
```

**Reply by bcakmakoglu** [reply]
A timeout shouldn't be necessary though.

Your `useEffect` dependencies don't seem consistent either.
You're using `flowInstance` in your `useEffect` but the dependency array lists `[options, fitView]`.
So either you're having two `fitView` functions in the same scope and using one as a dependency but calling another or the code you presented isn't what you're actually using 🤔


It should either be this

```ts
useEffect(() => {
	fitView()
}, [options, fitView])
```

or 
```ts
useEffect(() => {
	flowInstance.fitView()
}, [options, flowInstance])
```

Here's an example: https://codesandbox.io/p/sandbox/distracted-grass-kmm3hl?file=%2Fsrc%2FApp.tsx%3A34%2C11

**Reply by CSworenI** [reply]
@bcakmakoglu Sometimes it works and sometimes it doesn't, apparently. It's being inconsistent. I was able to get it to work with aligned dependencies and a 10ms delay, but not always.

**Reply by bcakmakoglu** [reply]
> For whatever reason, I have to mix them to get consistent fitView behavior.

I don't quite see why that would be necessary *but* I can understand the reasoning that a slight timeout might be necessary for `fitView` to trigger *after* the sidebar has been closed otherwise the `useEffect` would probably run before a re-render happened and the sidebar is closed.

If you're ok with using more or less arbitrary waiting times I'd say the solution is fine but I'd try to cleanup the dependencies to be consistent with the actual function you're using in the effect.

Like this: https://codesandbox.io/p/sandbox/distracted-sea-9vjlqr

**Reply by bcakmakoglu** [reply]
> Sometimes it works and sometimes it doesn't, apparently.

I suspect that what you're seeing or interpreting isn't quite what is happening behind the scenes.
As I said in the previous post, the issue you're having is most likely that
* When you open/close the sidebar, the dimensions of the viewport are re-calculated
* The dimensions are used when calling `fitView`
	* Even though the sidebar opens, `fitView` is triggered at a time where it probably uses the previously recorded dimensions instead of those with/without the sidebar, making it seem like fitting the view "failed"

At least that's how I interpret what I'm seeing.

**Reply by CSworenI** [reply]
@bcakmakoglu Yes, I believe that's exactly what's happening. When aligning the dependencies and removing the delay, it realigns to the previous state of the options pane. I'm finding that for a lot of what I'm doing (modals, sidebars, user-driven flow changes), I'm needing to use at least a tiny delay to see the changes appear at runtime, so I will continue applying this methodology. It's only the first draft of the application, so I'm OK with revisiting this perceived need during rewrite(s) for stability and performance.

Have you any suggestions for [resizer with sidecars](https://github.com/xyflow/xyflow/discussions/4534)? It is much needed functionality that seems to put our task trackers in the middle of the node as the length of the label grows.

**Reply by maxblank-stripe** [reply]
@CSworenI `setTimeout` works, but the graph is flickering because of the timeout. `react-flow-renderer` used to be able to run `fitView` immediately after the re-render, does `reactflow` renders anything asynchronously or why does it have to wait?
